content/documentation/internals/nats-server-protocol.md

```diff
@@ -0,0 +1,263 @@
+
+The NATS server implements a [zero allocation byte parser](https://youtu.be/ylRKac5kSOk?t=10m46s) that is fast and efficient.
+
+The NATS server clustering protocol is very similar to that of the NATS client protocol.  In the context of a cluster, it can be helpful to visualize a server being a proxy operating on behalf of its connected clients, subscribing, unsubscribing, sending and receiving messages.
+
+## NATS Server protocol conventions
+
+**Subject names and wildcards**: The NATS server protocol has the same features and restrictions as the client with respect to subject names and wildcards, with one addition - a server can publish subjects prefixed with `_SYS.`, which is reserved for messages that are internally generated by a server.
+
+**Field Delimiters**: The fields of NATS protocol messages are delimited by whitespace characters '` `' (space)  or `\t` (tab).
+Multiple whitespace characters will be treated as a single field delimiter.
+
+**Newlines**: Like other text-based protocols, NATS uses `CR` followed by `LF` (`CR+LF`, `\r\n`, `0x0D0A`) to terminate protocol messages.  This newline sequence is also used to mark the beginning of the actual message payload in a `PUB` or `MSG` protocol message.
+
+
+## NATS Server protocol messages
+
+The following table briefly describes the NATS cluster protocol messages.
+As in the client protocol, the NATS protocol operation names are case insensitive, thus `SUB foo 1\r\n` and `sub foo 1\r\n` are equivalent.
+
+Click the name to see more detailed information, including syntax:
+
+
+| OP Name              | Sent By          |    Description
+| -------------------- |:-----------------|:--------------------------------------------
+| [`INFO`](#INFO)      | All Servers      | Sent after initial TCP/IP connection and to update cluster knowledge
+| [`CONNECT`](#CONNECT)| All Servers      | Sent to establish a route
+| [`PUB`](#PUB)        | Sending Server   | Sent when a message is published by a client.
+| [`SUB`](#SUB)        | All Servers      | Subscribes to a subject on behalf of interested clients.
+| [`UNSUB`](#UNSUB)    | All Servers      | Unsubscribe (or auto-unsubscribe) from subject.
+| [`MSG`](#MSG)        | Receiving Server | Delivers a message payload to a subscriber on the local server.
+| [`PING`](#PINGPONG)  | All Servers      | PING keep-alive message
+| [`PONG`](#PINGPONG)  | All Servers      | PONG keep-alive response
+| [`+OK/-ERR`](#OKERR) | All Servers      | Indicates a protocol error. Will cause the server to disconnect.
+
+
+The following sections explain each protocol message.
+
+## <a name="INFO"></a>INFO
+
+### Syntax
+
+`INFO {["option_name":option_value],...}`
+
+The valid options are as follows:
+
+* `server_id`: The unique identifier of the NATS server
+* `version`: The version of the NATS server
+* `go`: The version of golang the NATS server was built with
+* `host`: The IP address of the NATS server host
```

Comment:
In the server's INFO, it is worth noting that `host` and `port` here refer to values specified by the `--cluster` command line parameter or `Options.ClusterHost` and `Options.ClusterPort`.